File: Island Perimeter.cpp
class Solution {
public:
    int getKey(int i, int j) {
        return (i << 10) | j;
    }

    pair<int, int> returnKey(int key) {
        int i = key >> 10;
        int j = ((1 << 10) - 1) & key;
        return make_pair(i, j);
    }

    int islandPerimeter(vector<vector<int>>& grid) {
        unordered_set<int> visited;
        int perimeter = 0;
        int n = grid.size(), m = grid[0].size();
        vector<pair<int, int>> q;

        bool to_break = false;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j]) {
                    q.push_back(make_pair(i, j));
                    to_break = true;
                    break;
                }
            }

            if (to_break) break;
        }

        while (q.size() > 0) {
            auto p = q.back();
            int i = p.first, j = p.second;
            int key = getKey(i, j);
            q.pop_back();

            if (visited.contains(key)) continue;

            visited.insert(key);
            vector<pair<int, int>> pairs{{i-1, j}, {i+1, j}, {i, j-1}, {i, j+1}};
            for (auto & xy: pairs) {
                int x = xy.first, y = xy.second;
                if (visited.contains(getKey(x, y))) continue;

                if (0 <= x && x < n && 0 <= y && y < m && grid[x][y]) {
                    q.push_back(xy);
                } else {
                    perimeter++;
                }
            }
        }
        return perimeter;
    }
};

File: Island Perimeter.py
class Solution:
    def islandPerimeter(self, grid: List[List[int]]) -> int:
        perimeter = 0
        n = len(grid)
        m = len(grid[0])

        for i in range(n):
            prev_occupied = 0
            for j in range(m):
                perimeter += grid[i][j] ^ prev_occupied
                prev_occupied = grid[i][j]
            perimeter += prev_occupied

        for j in range(m):
            prev_occupied = 0
            for i in range(n):
                perimeter += grid[i][j] ^ prev_occupied
                prev_occupied = grid[i][j]
            perimeter += prev_occupied
        return perimeter

# class Solution:
#     def islandPerimeter(self, grid: List[List[int]]) -> int:
#         visited = set()
#         perimeter = 0
#         n = len(grid)
#         m = len(grid[0])
#         q = []
#
#         to_break = False
#         for i in range(n):
#             for j in range(m):
#                 if grid[i][j]:
#                     q.append((i, j))
#                     to_break = True
#                     break
#             if to_break:
#                 break
#
#         while q:
#             i, j = q.pop()
#             if (i, j) in visited:
#                 continue
#
#             visited.add((i, j))
#             for (x, y) in [(i - 1, j), (i + 1, j), (i, j - 1), (i, j + 1)]:
#                 if (x, y) in visited:
#                     continue
#
#                 if 0 <= x < n and 0 <= y < m and grid[x][y]:
#                     q.append((x, y))
#                 else:
#                     perimeter += 1
#         return perimeter
